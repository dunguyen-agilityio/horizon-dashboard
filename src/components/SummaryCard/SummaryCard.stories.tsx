import type { Meta, StoryObj } from '@storybook/react';

import SummaryCard from '.';

import AddIcon from '@/icons/Add';

const meta = {
  title: 'Components/SummaryCard',
  component: SummaryCard,
  tags: ['autodocs'],
} satisfies Meta<typeof SummaryCard>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    title: 'Earnings',
    counter: 350.5,
    additionalContent: <AddIcon />,
  },
};

export const Secondary: Story = {
  args: {
    variant: 'reverse',
    title: 'Earnings',
    counter: 350.5,
    additionalContent: <AddIcon />,
  },
};

export const CustomAdditionalContent: Story = {
  args: {
    title: 'Earnings',
    counter: 350.5,
    additionalContent: <AddIcon />,
    additionalContentWrapper: 'bg-red-700 dark:bg-red-700',
  },
};

export const Increment: Story = {
  args: {
    variant: 'reverse',
    title: 'Earnings',
    counter: 350.5,
    trendValue: 20,
    trendDescription: 'since last month',
  },
};

export const Decrement: Story = {
  args: {
    variant: 'reverse',
    title: 'Earnings',
    counter: 350.5,
    trendValue: -20,
    trendDescription: 'since last month',
  },
};
