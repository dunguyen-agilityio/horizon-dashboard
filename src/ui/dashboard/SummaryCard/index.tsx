// Libs
import { cn } from '@nextui-org/theme';

// Components
import Text from '@/components/Text';

// Types
import { TEXT_SIZE, TEXT_VARIANT } from '@/types/text';

type LayoutVariant = 'primary' | 'secondary';

export type TrendInfo =
  | {
      trendValue: number;
      trendUnit?: string;
      trendDescription: string;
    }
  | { trendValue?: undefined };

export interface SummaryCardProps {
  variant?: LayoutVariant;
  additionalContent?: React.ReactElement;
  additionalContentWrapper?: string;
  title?: string;
  counter: number;
  unit?: string;
}

const classes = {
  primary: '',
  secondary: 'flex-row-reverse justify-between',
} as Record<LayoutVariant, string>;

const SummaryCard = ({
  variant = 'primary',
  additionalContent,
  counter,
  title,
  unit = '$',
  additionalContentWrapper,
  ...otherProps
}: SummaryCardProps & TrendInfo) => {
  const renderTrendContent = () => {
    if (otherProps.trendValue === undefined) return null;

    const { trendValue, trendDescription, trendUnit = '%' } = otherProps;

    const trendStyle = cn(
      'font-bold text-tiny',
      trendValue > 0 ? '!text-green-700' : '!text-red-700',
    );

    return (
      <Text className="text-tiny" variant={TEXT_VARIANT.SECONDARY}>
        <Text as="span" className={trendStyle} data-testid="additional-trend">
          {trendValue > 0 ? `+${trendValue}` : trendValue}
        </Text>
        <Text as="span" className={trendStyle}>
          {trendUnit}
        </Text>
        &nbsp;
        {trendDescription}
      </Text>
    );
  };

  return (
    <div
      className={cn(
        'flex gap-[18px] items-center bg-white dark:bg-indigo p-3 4xl:p-5 flex-1 rounded-md lg:min-w-[248px] 2xl:min-w-40',
        classes[variant],
      )}
    >
      {additionalContent && (
        <div
          className={cn(
            'h-14 min-w-14 rounded-full bg-gray dark:bg-indigo-light flex justify-center items-center',
            additionalContentWrapper
              ? `p-0 ${additionalContentWrapper}`
              : 'p-3',
          )}
        >
          {additionalContent}
        </div>
      )}
      <div className="flex flex-col gap-[3px]">
        <Text size={TEXT_SIZE.sm} variant={TEXT_VARIANT.SECONDARY} as="span">
          {title}
        </Text>
        <Text size={TEXT_SIZE.extra} as="b">
          &nbsp;
          {`${unit}${counter}`}
        </Text>
        {renderTrendContent()}
      </div>
    </div>
  );
};

export default SummaryCard;
