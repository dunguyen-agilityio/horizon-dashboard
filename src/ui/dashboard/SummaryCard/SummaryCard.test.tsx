import { render } from '@testing-library/react';

import SummaryCard from '.';
import AddIcon from '@/icons/Add';

describe('SummaryCard tests', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <SummaryCard
        title="Earnings"
        counter={350.5}
        additionalContent={<AddIcon />}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot', () => {
    const { container } = render(
      <SummaryCard
        title="Earnings"
        counter={350.5}
        additionalContent={<AddIcon />}
        additionalContentWrapper="bg-red-700 dark:bg-red-700"
        trendValue={20.3}
        trendDescription="since last month"
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Should match snapshot', () => {
    const { container } = render(
      <SummaryCard
        title="Earnings"
        counter={350.5}
        additionalContent={<AddIcon />}
        additionalContentWrapper="bg-red-700 dark:bg-red-700"
        trendValue={-9}
        trendDescription="since last month"
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
