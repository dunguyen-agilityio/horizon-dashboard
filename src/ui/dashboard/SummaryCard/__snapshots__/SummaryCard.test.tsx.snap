// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryCard tests Should match snapshot 1`] = `
<div>
  <div
    class="flex justify-between p-3 sm:justify-start sm:gap-[18px] items-center bg-white dark:bg-indigo 4xl:p-5 flex-1 rounded-md lg:min-w-[248px] 2xl:min-w-40"
  >
    <div
      class="h-14 min-w-14 rounded-full bg-gray dark:bg-indigo-light flex justify-center items-center p-3"
    >
      <svg
        class="[&>path]:stroke-primary [&>path]:dark:stroke-white"
        fill="none"
        height="17"
        viewBox="0 0 19 17"
        width="19"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.32703 3.71619V13.2787"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
        <path
          d="M14.3896 8.49744H4.26459"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
    <div
      class="flex flex-col gap-[3px]"
    >
      <span
        class="text-secondary dark:text-secondary text-small font-medium"
      >
        Earnings
      </span>
      <b
        class="text-primary dark:text-white text-1.5xl font-bold leading-xl"
      >
        Â 
        $350.5
      </b>
    </div>
  </div>
</div>
`;
